Use the site search endpoint for plugin search

gcsecsey opened the pull request:
### Changes proposed in this Pull Request

* Adds a new hook to use the `https://public-api.wordpress.com/rest/v1/sites/108986944/search` endpoint
* Adds the same parameters for author and category search as WPORG
* Moves some common code to utils and constants files

### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Visit `/plugins`
* Search for something generic (eg. `wordpress`)
* Verify that both WPCOM and WPORG results are loaded
* Verify that WPORG results have all the necessary information on the search result card (title, excerpt, rating, last update, number of installs)
* Scroll to the bottom, and verify that the search result list loads the next page of results 

**Videos**
Plugin search: https://d.pr/v/bvi2uz
Author search: https://d.pr/v/9ntpmF

### Known issues, next steps

#### Category search

The index contains only WPORG categories, and we can search for those. However, the categories used on the WPORG API (Popular, Featured), are stored in the `plugin_section` taxonomy and are not in the index response: 
https://github.com/WordPress/wordpress.org/blob/38e38a5b0bbca67affaf0ec70617db14f3f88f84/wordpress.org/public_html/wp-content/plugins/plugin-directory/class-plugin-directory.php#L195
https://github.com/WordPress/wordpress.org/blob/38e38a5b0bbca67affaf0ec70617db14f3f88f84/wordpress.org/public_html/wp-content/plugins/plugin-directory/class-plugin-directory.php#L760

We should assess if we need to add this taxonomy to the index as well.

#### Localization

The current implementation only supports results in english. The index stores all the translations in the MAG16, so we can create the logic of picking the correct keys in a follow-up PR.

#### Icons

The index contains icon file names in a serialized PHP array. These are deserialized and appended to the WPORG image host (`https://ps.w.org`). For eg. Woocommerce plugins, the PHP array is empty, so icon URLs can't be produced just from the index response. We'll need to find an alternative, more stable way of getting the icon URLs.

**To reproduce:** try searching for `WooTcsShipmentMaker`

| Search results (index) | Plugin details page (WPORG) |
| ------ | ----- |
| ![image](https://user-images.githubusercontent.com/11555574/165253817-dfc5ab8b-1394-4789-a717-4d4692a6ffec.png) | ![image](https://user-images.githubusercontent.com/11555574/165253859-668568c0-fb22-43cb-a670-d2da8c6f9630.png) |

#### Author profile link

The index response doesn't contain the `author_profile` URL, which is a WPORG profile URL. I'm not sure we need it though, as the author name currently redirects to the author search page, and not the `author_profile` URL. It's also only used on the plugin details page, not the search results page.

| Search results | Plugin details page |
| ------ | ----- |
| ![image](https://user-images.githubusercontent.com/11555574/165255359-e727067b-ba8b-484a-844e-1e640cc8110e.png) | ![image](https://user-images.githubusercontent.com/11555574/165255640-169b2754-5198-48b3-a65a-f0620c3d832a.png) |

#### Plugin details page

We can query for an exact match on the product slug from the index, which would let us get the same data to display on the plugin details page. However, as the plugin details page uses more fields from the WPORG response, we should investigate in a follow-up issue whether we need all the required fields to display the plugin page.

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #62729

gcsecsey committed:
WIP update WPOrg hook to use site search endpoint

gcsecsey committed:
WIP update parameters

gcsecsey committed:
WIP map index fields to wporg plugins endpoint results

gcsecsey committed:
WIP update mapping, add request

gcsecsey committed:
WIP create site search hook

gcsecsey committed:
add phpunserialize, revert minor changes

gcsecsey committed:
WIP udpate types, convert query to infinitequery

gcsecsey committed:
WIP pass response data to infinite query

gcsecsey committed:
fix pagination

lsl commented:
Will this somehow make use of this new flag? 12392-gh-Automattic/woocommerce.com#issuecomment-1105140159

gcsecsey committed:
fix infinitequery fetching all pages on load

Cache key containing the page param would cause an automatic refetch of the next page

gcsecsey committed:
use cache key util in WPORG requests

gcsecsey committed:
use site search hook on plugins-browser page

gcsecsey committed:
Merge branch 'trunk' into update/use-es-for-plugin-search

gcsecsey committed:
add tag to plugin query options

gcsecsey commented:
> Will this somehow make use of this new flag? [12392-gh-Automattic/woocommerce.com#issuecomment-1105140159](https://href.li/?12392-gh-Automattic/woocommerce.com#issuecomment-1105140159)

@lsl the MVP version only supports WPORG results, so I don't think we'll be able to filter for `marketplace_categories` just yet. D79130-code is adding WPCOM and WCCOM results to a new index and would let us filter based on taxonomies.

gcsecsey committed:
update tests

gcsecsey committed:
update unserialize lib

change the dependency to one that has exported types, and move the dependency under the client sub-package

gcsecsey committed:
add type declaration to unserialize lib

WBerredo commented:
**Author profile link** -> currently this we define the profile keyword this way: https://github.com/Automattic/wp-calypso/blob/6406aceea9155168d38ab47ba7c18acbaa346678/client/lib/plugins/utils.js#L284-L288

WBerredo commented on client/data/marketplace/constants.ts:
```diff
@@ -0,0 +1,10 @@
+export const DEFAULT_FIRST_PAGE = 1;
+export const DEFAULT_CATEGORY = 'all';
+
+export const SITE_SEARCH_ENDPOINT =
+	'https://public-api.wordpress.com/rest/v1/sites/108986944/search';
```
Any chance you add a comment or a different constant explaining this value `108986944`?

gcsecsey replied:
Good idea to make it more clear, I've added it as a constant! 🙂

WBerredo commented on client/data/marketplace/use-site-search-es-query.ts:
```diff
@@ -0,0 +1,217 @@
+	};
+	const icons: Record< string, Icon > = phpUnserialize( iconsArray );
+	return Object.values( icons ).reduce(
+		( prev: Record< string, string >, { resolution, filename, location, revision } ) => {
+			prev[
+				resolution
+			] = `https://ps.w.org/${ pluginSlug }/${ location }/${ filename }?rev=${ revision }`;
+			return prev;
+		},
```
Will we always have those 4 parameters? (resolution, filename, location, revision)

Do you know if there is any kind of fallback for the times we don't have some of them?

gcsecsey replied:
The icons array contains these keys after converted to JSON: 

```json
{
  "icon-128x128.png": 
    "filename": "icon-128x128.png"
    "locale": ""
    "location": "assets"
    "resolution": "128x128"
    "revision": 2366418
}
```

So far, all results contained those 4 keys, and I couldn't find any results where the `locale` key is used. As this is originally stored as a PHP array, I wasn't able to filter for it in ES.

I don't know what could be a sensible fallback for these, as the index doesn't contain anything else regarding icons. For now, the `reduce` returns an empty array if there are no icons in the `assets` folder, and the default icon is used on the UI.

WBerredo replied:
At first, I was thinking about two possible fallbacks:
* resolution: but we should be handling it later down the line when we choose which version to pick
* revision: I was wondering it was sequential like `1, 2, 3, etc`, so we could use a fallback to `1`. Thanks for adding a response example here as I was not guessing it right.

That said, I think the current version is fine. Ty.

gcsecsey replied:
For the plugins that have icon assets uploaded to WPORG, these fields are always indexed. We have to have a fallback for plugins that don't have icons uploaded, I've just found out that WPORG generates a pattern for these (eg. [here](https://wordpress.org/plugins/slideshare/)). I'll incorporate a similar fallback in a follow-up PR.

WBerredo commented on client/data/marketplace/use-site-search-es-query.ts:
```diff
@@ -0,0 +1,217 @@
+	);
+};
+
+const createAuthorUrl = ( header_author: string, header_author_uri: string ) =>
+	`<a href="${ header_author_uri }">${ header_author }</a>`;
```
Nitpicking: I would use camelCase when we are not explicitly using a fiedName.

```suggestion
const createAuthorUrl = ( headerAuthor: string, headerAuthorUri: string ) =>
	`<a href="${ headerAuthorUri }">${ headerAuthor }</a>`;
```

WBerredo commented on client/data/marketplace/use-site-search-es-query.ts:
```diff
@@ -0,0 +1,217 @@
+			name: hit.title_en, // TODO: add localization
+			slug: hit.slug,
+			version: hit.stable_tag,
+			author: createAuthorUrl(
+				hit[ 'meta.header_author.value' ],
+				hit[ 'meta.header_author_uri.value' ]
+			),
```
Do we have a `contributors` field? 

We saw we were getting better results when using it to find the author for search.

Reference: https://github.com/Automattic/wp-calypso/pull/62777#issuecomment-1109980813

gcsecsey replied:
I've tried using the function, or implementing it again for the ES index. The index has something similar, the problem is that the data is in a different structure:

Each hit contains a `taxonomy.plugin_contributors` and a flat `contributors` array:
```json
{
    "taxonomy": {
        "plugin_contributors": [
            {
                "name": "automattic",
                "slug": "automattic",
                "term_id": "77494"
            },
            {
                "name": "mikejolley",
                "slug": "mikejolley",
                "term_id": "77602"
            }
        ]
    },
    "contributors": [
        "automattic",
        "mikejolley"
    ]
}
```

Also, the `meta.header_author` object contains a display name, while the rest of the fields contain the slug/username. In the case of Automattic it's just an upper/lower case difference, but there are authors for which these two are entirely different. Without a slug field to match on, it's hard to choose the correct contributor as the author from the list.
```json
"header_author": {
    "value": [
        "Automattic"
    ],
    "long": [
        0
    ],
    "double": [
        0
    ],
    "boolean": [
        true
    ]
},
```

I'll look into how the contributors array is ordered, maybe the header author is always the first contributor.

WBerredo commented on client/data/marketplace/use-site-search-es-query.ts:
```diff
@@ -0,0 +1,217 @@
+	throw new Error( await response.text() );
+}
+
+export async function fetchWPOrgPluginsFromSiteSearch( options: PluginQueryOptions ) {
```
Nitpicking

The ES index will have only WPOrg plugins for now?

Maybe it is a good idea to use a more generalized method name as it should come from the same place(unified index) in the future, and terms as WPcom and WPorg plugins won't make sense anymore.

gcsecsey replied:
AFAIK yes, until D79130-code is landed we won't be able to query other sources.

I agreee that the name should be more generic, I've updated it. 👍

WBerredo commented on client/data/marketplace/use-site-search-es-query.ts:
```diff
@@ -0,0 +1,217 @@
+
+const extractPages = ( pages: Array< { hits: ESHits } > = [] ) => {
+	const allHits = pages.flatMap( ( page ) => page.hits );
+	return normalizePluginsList( mapIndexResultsToPluginData( allHits ) );
```
Nitpicking: I think we can simplify this piece in the future. We are mapping the ES results to look like the old requests and then we normalize this data for usage in the application. I would say we could do this straight from ES results to application-ready data.

That said, I think it should be a follow-up PR as we are changing a lot already at this one.

gcsecsey replied:
I agree that the mapping could be further simplified, but I wanted to touch as few layers as possible with the MVP. My original idea was to replace only the fetch layer within the react-query and do a data-mapping that would let the `select` pick up the response the same way. Unfortunately, the responses are so different (e.g pagination for the ES response has to be calculated from the query params and the `total` number in the response), that I had to create a new query after all.

Later, when all results can be queried from the same index, we can definitely come up with a simpler mapping and remove a lot of redundant logic.

WBerredo reviewed (dismissed):
Before merging this PR, we must discuss if we will add an experiment to compare the results of both experiences. 

In a positive case, we would need some work setting up this experiment and some small changes to `PluginsBrowser` component.

cpapazoglou reviewed (dismissed):
Thanks for working on this @gcsecsey! I works quite good and the API response is 25% faster 👏👏 👏 

Some feedback:
1. Let's use a feature flag and conditionally (manually) enable it in all environments. This will allow us to push the infrastructure changes and then iterate on the details.
2. Ratings need some different parsing as it seems
![CleanShot 2022-04-27 at 11 05 03@2x](https://user-images.githubusercontent.com/12430020/165471743-154ff82f-13d6-4429-8843-ffded0c80834.png)
3. We will definitely need handle author search better as @WBerredo suggested
4. It seems that sorting is a tad different 

|Before | After|
|-------|------|
|![](https://user-images.githubusercontent.com/12430020/165472258-74642600-594b-4a1d-8f37-aeb2da6f7339.png)|![](https://user-images.githubusercontent.com/12430020/165472208-bf5cf780-c912-4c6f-84ae-fd3c595ef73a.png)|

I guess this is `active_installations` related. Can we sort by this property?

gcsecsey committed:
PR feedback

gcsecsey committed:
PR feedback

gcsecsey committed:
map 5 star rating to percentage

gcsecsey committed:
score results similarly to WPORG

gcsecsey commented on client/data/marketplace/use-site-search-es-query.ts:
```diff
@@ -0,0 +1,300 @@
+						missing: 1,
+					},
+				},
+				{
+					filter: {
+						range: {
+							active_installs: {
+								lte: 1000000,
+							},
+						},
+					},
+					exp: {
+						active_installs: {
+							origin: 1000000,
+							offset: 0,
+							scale: 900000,
+							decay: 0.75,
+						},
+					},
+				},
+				{
+					field_value_factor: {
+						field: 'support_threads_resolved',
+						factor: 0.25,
+						modifier: 'log2p',
+						missing: 0.5,
+					},
+				},
+				{
+					field_value_factor: {
+						field: 'rating',
+						factor: 0.25,
+						modifier: 'sqrt',
+						missing: 2.5,
+					},
+				},
+			],
+			score_mode: 'multiply',
+		},
+	};
```
I've tried to port the sorting criteria that is sent to the site search endpoint from the WPORG backend here: https://github.com/WordPress/wordpress.org/blob/7f3e0d50e02008f5edd46d281052caf827661550/wordpress.org/public_html/wp-content/plugins/plugin-directory/libs/site-search/jetpack-search.php#L1040

It still doesn't match the original results, but results with a high number of installations and recent updates are given higher scores.

| Before | After | WPORG API |
| ------ | ----- | --- | 
| ![image](https://user-images.githubusercontent.com/11555574/165739180-0f208cfb-dbb8-425e-8429-4933e1537908.png) | ![image](https://user-images.githubusercontent.com/11555574/165738985-7a9d893c-33b7-40c7-81ea-de0bcfb41524.png) | ![image](https://user-images.githubusercontent.com/11555574/165739038-81f71ac6-266f-47af-90f9-278979857392.png) |

gcsecsey committed:
add `marketplace-jetpack-plugin-search` feature flag

gcsecsey committed:
add feature flag check to plugins browser

gcsecsey commented:
I've added the `marketplace-jetpack-plugin-search` feature flag, it is disabled by default on all environments.

To use the site index search endpoint instead of the WPORG API when searching, the feature flag can be enabled by:
- running `ENABLE_FEATURES=marketplace-jetpack-plugin-search yarn start`
- adding `?flags=marketplace-jetpack-plugin-search` at the end of the URL

gcsecsey commented:
I'll move all blockers identified so far to [the issue of this PR](https://github.com/Automattic/wp-calypso/issues/62729). As this feature is now behind a feature flag, we can merge and deploy this PR.

cpapazoglou reviewed (approved):
- ✅ Uses the default method (.org API) when not feature flag is added.
- ✅ Uses the elastic search when `?flags=marketplace-jetpack-plugin-search` is added
- ✅ Feature is by default false in all environments.

Good work @gcsecsey 👏 ! Let's 🚢  this beast and iterate on it.
